File: src/main/java/com/project/pantau/controller/AuthController.java
package com.project.pantau.controller;

import com.project.pantau.common.response.ApiResponse;
import com.project.pantau.dto.auth.AuthResponse;
import com.project.pantau.dto.auth.LoginRequest;
import com.project.pantau.dto.auth.RegisterRequest;
import com.project.pantau.service.AuthService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/auth")
@RequiredArgsConstructor
public class AuthController {
    private final AuthService authService;

    @PostMapping("/register")
    public ResponseEntity<ApiResponse<AuthResponse>> register(
            @Valid @RequestBody RegisterRequest request
    ) {
        AuthResponse response = authService.register(request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(new ApiResponse<>(
                        true,
                        "Registration successful",
                        response
                ));
    }

    @PostMapping("/login")
    public ResponseEntity<ApiResponse<AuthResponse>> login(
            @Valid @RequestBody LoginRequest request
    ) {
        AuthResponse response = authService.login(request);
        return ResponseEntity.ok(new ApiResponse<>(
                true,
                "Login successful",
                response
        ));
    }
}

File: internal/controller/auth_controller.go
package controller

import (
	"log/slog"
	"pantau/internal/dto/auth"
	"pantau/internal/service"
	"pantau/pkg/response"

	"github.com/gofiber/fiber/v3"
)

type AuthController interface {
	Register(ctx fiber.Ctx) error
	Login(ctx fiber.Ctx) error
}

type authControllerImpl struct {
	authService service.AuthService
}

func NewAuthController(
	authService service.AuthService,
) AuthController {
	return &authControllerImpl{authService: authService}
}

func (controller *authControllerImpl) Login(ctx fiber.Ctx) error {
	var req auth.LoginRequest

	if err := ctx.Bind().Body(&req); err != nil {
		slog.Error("[AuthController] Failed to bind request", "email", req.Email, "error", err)
		return err
	}

	result, err := controller.authService.Login(
		ctx.Context(),
		req,
	)
	if err != nil {
		slog.Error("[AuthController] Failed to login", "email", req.Email, "error", err)
		return err
	}

	return ctx.
		Status(fiber.StatusOK).
		JSON(response.Success(result))
}

func (controller *authControllerImpl) Register(ctx fiber.Ctx) error {
	var req auth.RegisterRequest

	if err := ctx.Bind().Body(&req); err != nil {
		slog.Error("[AuthController] Failed to bind request", "email", req.Email, "error", err)
		return err
	}

	result, err := controller.authService.Register(
		ctx.Context(),
		req,
	)
	if err != nil {
		slog.Error("[AuthController] Failed to register", "email", req.Email, "error", err)
		return err
	}

	return ctx.
		Status(fiber.StatusCreated).
		JSON(response.Success(result))
}
